sarc/jobs/series.py

```diff
@@ -218,6 +220,133 @@ def compute_job_statistics(job: SlurmJob):
+
+    total = count_jobs(**jobs_args)
+
+    rows = []
+    # Fetch all jobs from the clusters
+    for job in tqdm(get_jobs(**jobs_args), total=total, desc="load job series"):
+        if job.end_time is None:
+            job.end_time = datetime.now(tz=MTL)
+
+        # For some reason start time is not reliable, often equal to submit time,
+        # so we infer it based on end_time and elapsed_time.
+        job.start_time = job.end_time - timedelta(seconds=job.elapsed_time)
+
+        unclipped_start = None
+        unclipped_end = None
+        if clip_time:
+            assert start is not None
+            assert end is not None
+            # Clip the job to the time range we are interested in.
+            # NOTE: This should perhaps be helper function for dataframes so that we don't force clipping raw data
+            #       during loading.
+            unclipped_start = job.start_time
+            job.start_time = max(job.start_time, start)
+            unclipped_end = job.end_time
+            job.end_time = min(job.end_time, end)
+            # Could be negative if job started after end. We don't want to filter
+            # them out because they have been submitted before end, and we want to
+            # compute their wait time.
+            job.elapsed_time = max((job.end_time - job.start_time).total_seconds(), 0)
+
+        if job.stored_statistics is None:
+            job_series = DUMMY_STATS.copy()
+        else:
+            job_series = job.stored_statistics.dict()
+            job_series = {k: select_stat(k, v) for k, v in job_series.items()}
+
+        # TODO: Why is it possible to have billing smaller than gres_gpu???
+        billing = job.allocated.billing or 0
+        gres_gpu = job.allocated.gres_gpu or 0
+        if gres_gpu:
+            job_series["gpu_allocated"] = max(billing, gres_gpu)
+            job_series["cpu"] = job.allocated.cpu
+        else:
+            job_series["gpu_allocated"] = 0
+            job_series["cpu"] = (
+                max(billing, job.allocated.cpu) if job.allocated.cpu else 0
+            )
+        job_series["gpu_requested"] = gres_gpu
+        job_series["duration"] = job.elapsed_time
+        job_series["mem"] = job.allocated.mem
+        job_series["submit"] = job.submit_time
+        job_series["start"] = job.start_time
+        job_series["end"] = job.end_time
+        if clip_time:
+            job_series["unclipped_start"] = unclipped_start
+            job_series["unclipped_end"] = unclipped_end
+
+        job_series.update(job.dict())
+        rows.append(job_series)
+
+    if fields is not None:
+        if isinstance(fields, list):
+            rows = [{key: job_dict[key] for key in fields} for job_dict in rows]
+        else:
+            assert isinstance(fields, dict)
+            rows = [
+                {new_name: job_dict[old_name] for old_name, new_name in fields.items()}
+                for job_dict in rows
+            ]
+
+    if callback:
```

Comment:
Le callback devrait être appelé à la fin de la boucle-for. Il ne sert pas à grand chose sinon.

Reply:
Donc, ça veut dire que le callback sera appelé à chaque tour de boucle, et recevra les `rows` augmentées d'un job par tour ? Ça fera beaucoup d'appels de callbacks. Est-ce le fonctionnement souhaité ?

Reply:
C'est pour pouvoir utiliser le callback comme fonction de checkpoint ou autre chose. Cette fonction peut prendre plus de 30 minutes à exécuter si on query un an de données.

Reply:
Ok ! Fait ! https://github.com/mila-iqia/SARC/pull/89/commits/95f925c6b9403d40700e074c361e83fcedcec329